### PR TITLE
[3d] Measurement tool

### PR DESCRIPTION
**Pull request by ismailsunni**
## Description
Include a few sentences describing the overall goals for this PR (pull request). If applicable also add screenshots.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/app/3d/qgs3dmaptoolmeasureline.cpp [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+{
+  if ( event->button() == Qt3DRender::QPickEvent::LeftButton )
+  {
+    if ( mMeasureLineTool->done() )
+    {
+      mMeasureLineTool->mDialog->restart();
+    }
+
+    // Left button, keep addinng point
+    qInfo() << "handlePickOnVectorLayer" ;
```
Comment (wonder-sk) [inline src/app/3d/qgs3dmaptoolmeasureline.cpp]: If these are to stay, it's better to use QgsDebugMsg because the logs can be enabled/disabled. If the qInfo()/qDebug() are just for debugging, best to remove them.

File: src/app/3d/qgs3dmaptoolmeasureline.cpp [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+{
+  public:
+    Qgs3DMapToolMeasureLinePickHandler( Qgs3DMapToolMeasureLine *measureLineTool ): mMeasureLineTool( measureLineTool ) {}
+    void handlePickOnVectorLayer( QgsVectorLayer *vlayer, QgsFeatureId id, const QVector3D &worldIntersection, Qt3DRender::QPickEvent *event ) override;
+  private:
+    Qgs3DMapToolMeasureLine *mMeasureLineTool = nullptr;
+};
+
+void Qgs3DMapToolMeasureLinePickHandler::handlePickOnVectorLayer( QgsVectorLayer *vlayer, QgsFeatureId id, const QVector3D &worldIntersection, Qt3DRender::QPickEvent *event )
+{
```
Comment (wonder-sk) [inline src/app/3d/qgs3dmaptoolmeasureline.cpp]: looks like this function is mostly a copy of the content in onTerrainPicked() - would it be possible to merge the two to avoid code duplication?

File: src/app/3d/qgs3dmaptoolmeasureline.cpp [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+  connect( mCanvas->scene(), &Qgs3DMapScene::terrainEntityChanged, this, &Qgs3DMapToolMeasureLine::onTerrainEntityChanged );
+  mPickHandler.reset( new Qgs3DMapToolMeasureLinePickHandler( this ) );
+
+  // Line style
+  mLineSymbol = new QgsLine3DSymbol;
+  mLineSymbol->setRenderAsSimpleLines( true );
+  mLineSymbol->setWidth( 4 );
+  QgsPhongMaterialSettings phongMaterial;
+  phongMaterial.setAmbient( Qt::yellow );
+  phongMaterial.setDiffuse( Qt::green );
```
Comment (wonder-sk) [inline src/app/3d/qgs3dmaptoolmeasureline.cpp]: I think the diffuse color is ignored for the "simple" lines - only ambient color is used

File: src/app/3d/qgs3dmaptoolmeasureline.cpp [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+  qInfo() << "Constructed";
+  connect( mCanvas->scene(), &Qgs3DMapScene::terrainEntityChanged, this, &Qgs3DMapToolMeasureLine::onTerrainEntityChanged );
+  mPickHandler.reset( new Qgs3DMapToolMeasureLinePickHandler( this ) );
+
+  // Line style
+  mLineSymbol = new QgsLine3DSymbol;
+  mLineSymbol->setRenderAsSimpleLines( true );
+  mLineSymbol->setWidth( 4 );
+  QgsPhongMaterialSettings phongMaterial;
+  phongMaterial.setAmbient( Qt::yellow );
```
Comment (wonder-sk) [inline src/app/3d/qgs3dmaptoolmeasureline.cpp]: Shall we use the same color as the 2D measure tool rubber band's color?

File: src/app/3d/qgs3dmaptoolmeasureline.cpp [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+  : Qgs3DMapTool( canvas )
+{
+  qInfo() << "Constructed";
+  connect( mCanvas->scene(), &Qgs3DMapScene::terrainEntityChanged, this, &Qgs3DMapToolMeasureLine::onTerrainEntityChanged );
+  mPickHandler.reset( new Qgs3DMapToolMeasureLinePickHandler( this ) );
+
+  // Line style
+  mLineSymbol = new QgsLine3DSymbol;
+  mLineSymbol->setRenderAsSimpleLines( true );
+  mLineSymbol->setWidth( 4 );
```
Comment (wonder-sk) [inline src/app/3d/qgs3dmaptoolmeasureline.cpp]: As discussed in our chat today, would be good to use Absolute altitude clamping so that the terrain is not added twice.

File: src/app/3d/qgs3dmaptoolmeasureline.cpp [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+
+  mCanvas->scene()->registerPickHandler( mPickHandler.get() );
+
+  // Initialize the measurement line
+  mMeasurementLine = new QgsLineString();
+  mMeasurementLine->addZValue();
+
+  // Initialize measurement feature
+  mMeasurementFeature = new QgsFeature( QgsFeatureId( 1 ) );
+  mMeasurementFeature->setGeometry( QgsGeometry( mMeasurementLine ) );
```
Comment (wonder-sk) [inline src/app/3d/qgs3dmaptoolmeasureline.cpp]: I think we do not need mMeasurementLine and mMeasurementFeature as member variables. They could be just created as local variables.

We also need to take care of object ownership:
- mMeasurementLine gets owned by QgsGeometry, so it is fragile to actually keep long-term pointer to it (may cause crash if the geometry is replaced which may cause premature destruction of the linestring and then crashes
- mMeasurementFeature leaks - it is just copied in addFeature()

File: src/app/3d/qgs3dmaptoolmeasureline.cpp [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+
+  // Initialize the line layer
+  QString mapCRS = mCanvas->map()->crs().authid();
+  mMeasurementLayer = new QgsVectorLayer( QStringLiteral( "LineStringZ?crs=" ) + mapCRS, QStringLiteral( "Measurement" ), QStringLiteral( "memory" ) );
```
Comment (wonder-sk) [inline src/app/3d/qgs3dmaptoolmeasureline.cpp]: maybe better to create the measurement layer only when the tool is activated for the first time. Subsequent activations could just "reset" the feature geometry like in restart(). Otherwise we may be leaving some memory leaks of vector layers after each activation.

File: src/app/3d/qgs3dmaptoolmeasureline.cpp [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+  QgsPoint addedPoint( point );
+  qInfo() << "Is 3D point? " << addedPoint.is3D();
+
+  // Append point
+  mPoints.append( addedPoint );
+
+  mMeasurementLine->addVertex( addedPoint );
+
+  mMeasurementLayer->startEditing();
+  QgsGeometry *newMeasurementLine = new QgsGeometry( mMeasurementLine );
```
Comment (wonder-sk) [inline src/app/3d/qgs3dmaptoolmeasureline.cpp]: a simpler solution than modifying mMeasurementLine would be to just create a new QgsLineString from mPoints whenever we need to call changeGeometry() - like that we only keep the list of points in one place.

File: src/app/3d/qgs3dmaptoolmeasureline.cpp [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+  mPoints.clear();
+
+  mDone = true;
+  mMeasurementLine->clear();
+
+  mMeasurementLayer->startEditing();
+  QgsGeometry *newMeasurementLine = new QgsGeometry( mMeasurementLine );
+  qInfo() << "Current line: " << newMeasurementLine->asWkt();
+  mMeasurementLayer->changeGeometry( mMeasurementFeature->id(), *newMeasurementLine );
+  mMeasurementLayer->commitChanges();
```
Comment (wonder-sk) [inline src/app/3d/qgs3dmaptoolmeasureline.cpp]: It would be good to have a single function e.g. updateMeasurementLayer() that would change geometry and commit changes based on mPoints - and multiple functions can use it do de-duplicate code.

File: src/app/3d/qgs3dmaptoolmeasureline.h [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    //! Removes the last point
+    void undo();
+
+    //! Returns reference to array of the points
+    QVector<QgsPoint> points() const;
+
+    // Inherited from Qgs3DMapTool
+    void mousePressEvent( QMouseEvent *event ) override { Q_UNUSED( event )}
+    void mouseReleaseEvent( QMouseEvent *event ) override { Q_UNUSED( event )}
+    void mouseMoveEvent( QMouseEvent *event ) override {Q_UNUSED( event )}
```
Comment (wonder-sk) [inline src/app/3d/qgs3dmaptoolmeasureline.h]: If you do not change anything in the implementation of mouse*Event() functions, then it's better to leave them out (they have empty implementation in the base class.

File: src/app/3d/qgs3dmeasuredialog.cpp [hunk 11]
```diff
@@ -0,0 +1,385 @@
...
+
+  repopulateComboBoxUnits();
+
+  // Choose unit
+  if ( mUseMapUnits )
+    mUnitsCombo->setCurrentIndex( mUnitsCombo->findData( QgsUnitTypes::DistanceUnknownUnit ) );
+  else
+    mUnitsCombo->setCurrentIndex( mUnitsCombo->findData( QgsProject::instance()->distanceUnits() ) );
+
+  QgsMapCanvas *canvas2D = QgisApp::instance()->mapCanvas();
```
Comment (wonder-sk) [inline src/app/3d/qgs3dmeasuredialog.cpp]: I think we should not work with 2D map canvas here - only focus on 3D map canvas - it has its own CRS.

Also elsewhere in this dialog I think we should not deal with 2D map canvas.

File: src/app/qgsoptions.cpp [hunk 12]
```diff
@@ -553,6 +553,53 @@ QgsOptions::QgsOptions( QWidget *parent, Qt::WindowFlags fl, const QList<QgsOpti
     mKeepBaseUnitCheckBox->setChecked( false );
   }
 
+  // ***** 3D measurement settings *****
```
Comment (wonder-sk) [inline src/app/qgsoptions.cpp]: Hmm do we need to have special configuration of decimal places, units, colors for 3D measure tool? I would suggest to reuse the 2D measure tool's configuration. It will keep the options simpler and it is also probably what people expect (not to have two different sets of options for measure tool)

File: src/core/qgsdistancearea.cpp [hunk 13]
```diff
@@ -347,6 +347,59 @@ double QgsDistanceArea::measureLine( const QgsPointXY &p1, const QgsPointXY &p2
...
+  double result;
+
+  try
+  {
+    QgsPoint pp1 = p1, pp2 = p2;
+
+    QgsDebugMsgLevel( QStringLiteral( "Cartesian calculation on canvas coordinates" ), 4 );
+    result = p2.distance3D( p1 );
+  }
+  catch ( QgsCsException &cse )
```
Comment (wonder-sk) [inline src/core/qgsdistancearea.cpp]: We don't do any CRS reprojection, so we don't need to catch exceptions here... (nor in the other measureLine3D())

**Review by wonder-sk — COMMENTED** [review]
Good stuff - I have highlighted a bunch of things worth addressing...

File: src/app/3d/qgs3dmeasuredialog.cpp (hunk 11 above)
Comment (lbartoletti) [reply src/app/3d/qgs3dmeasuredialog.cpp]: I agree, it should be separated. But, if possible, it would be nice to get 3D length on 2D map canvas too (and fix https://github.com/qgis/QGIS/issues/25576)

Comment (wonder-sk) [reply src/app/3d/qgs3dmeasuredialog.cpp]: Handling 3D length on 2D canvas is much more tricky I think. In 2D canvas the user may not be aware that the lengths could be calculated in 3D and that Z coordinates may (or may not) be picked from the snapped points. On the other hand in 3D canvas it is easier because it is clear that measurement points are from 3D world _and_ the Z values are consistently applied (some layers may have relative Z values, other may have absolute Z values - all that is made consistent in the 3D scene config and 3D renderer config of individual layers).

File: python/core/auto_generated/qgsdistancearea.sip.in [hunk 14]
```diff
@@ -224,6 +224,31 @@ Measures the distance between two points.
 
 :return: distance between points. The units for the returned distance can be retrieved by calling lengthUnits().
 
+.. seealso:: :py:func:`lengthUnits`
+%End
+
+    double measureLine3D( const QVector<QgsPoint> &points ) const;
+%Docstring
+Measures the length of a line with multiple segments in 3D.
+Support cartesian calculation only.
```
Comment (nyalldawson) [inline python/core/auto_generated/qgsdistancearea.sip.in]: I don't think this belongs in QgsDistanceArea if it has no handling of ellipsoidal/geodesic calculations. Rather it belongs in the geometry classes themselves.

See https://github.com/qgis/qgis4.0_api/issues/90 - this class should actually be named "QgsGeodesicUtils".

File: src/3d/qgs3dmapscene.cpp [hunk 15]
```diff
@@ -52,6 +52,8 @@
 
 #include "qgslinematerial_p.h"
 
+#include <qdebug.h>
```
Comment (nyalldawson) [inline src/3d/qgs3dmapscene.cpp]: Needed?

File: src/app/3d/qgs3dmaptoolmeasureline.cpp [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+  // Add feature to layer
+  mMeasurementLayer->startEditing();
+  mMeasurementLayer->addFeature( *mMeasurementFeature );
+  mMeasurementLayer->commitChanges();
+
+  // Set style
+  setMeasurementLayerRenderer( mMeasurementLayer );
+
+  // Add layer to canvas
+  qInfo() << "Current layer: " << mCanvas->map()->layers();
```
Comment (nyalldawson) [inline src/app/3d/qgs3dmaptoolmeasureline.cpp]: Remove in final version

File: src/app/3d/qgs3dmaptoolmeasureline.cpp [hunk 17]
```diff
@@ -0,0 +1,304 @@
...
+  //mCanvas->map()->setLayers( mCanvas->map()->layers() << mMeasurementLayer );
+  mCanvas->map()->setRenderers( QList<QgsAbstract3DRenderer *>() << mMeasurementLayer->renderer3D()->clone() );
+
+  qInfo() << "Current layer after adding: " << mCanvas->map()->layers();
```
Comment (nyalldawson) [inline src/app/3d/qgs3dmaptoolmeasureline.cpp]: Remove all of these

File: src/app/3d/qgs3dmaptoolmeasureline.cpp [hunk 18]
```diff
@@ -0,0 +1,304 @@
...
+    // Finish measurement
+    qInfo() << "Finish measurement";
+    mDone = true;
+  }
+  else if ( event->button() == Qt3DRender::QPickEvent::MiddleButton )
+  {
+    qInfo() << "Undo - middle button";
+    undo();
+  }
+
```
Comment (nyalldawson) [inline src/app/3d/qgs3dmaptoolmeasureline.cpp]: Cleanup these empty lines for final version

File: src/app/3d/qgs3dmaptoolmeasureline.cpp [hunk 19]
```diff
@@ -0,0 +1,304 @@
...
+}
+
+void Qgs3DMapToolMeasureLine::restart()
+{
+  mPoints.clear();
+
+  mDone = true;
+  mMeasurementLine->clear();
+
+  mMeasurementLayer->startEditing();
```
Comment (nyalldawson) [inline src/app/3d/qgs3dmaptoolmeasureline.cpp]: Seems redundant to use the vector layer edit buffer here -- maybe just change directly on the underlying provider and avoid the added expense?

File: src/app/3d/qgs3dmeasuredialog.cpp [hunk 20]
```diff
@@ -0,0 +1,385 @@
...
+  qInfo() << "Last line " << lastPoint.x() << lastPoint.y() << lastPoint.z();
+  qInfo() << "Last line " << secondLastPoint.x() << secondLastPoint.y() << secondLastPoint.z();
+  return qSqrt(
+           ( lastPoint.x() - secondLastPoint.x() ) * ( lastPoint.x() - secondLastPoint.x() ) +
+           ( lastPoint.y() - secondLastPoint.y() ) * ( lastPoint.y() - secondLastPoint.y() ) +
+           ( lastPoint.z() - secondLastPoint.z() ) * ( lastPoint.z() - secondLastPoint.z() )
+         );
+}
+
+void Qgs3DMeasureDialog::updateUi()
```
Comment (nyalldawson) [inline src/app/3d/qgs3dmeasuredialog.cpp]: This is a LOT of duplicated logic from the 2d case -- can we refactor this to avoid the added tech debt?

File: src/app/qgsoptions.cpp (hunk 12 above)
Comment (nyalldawson) [reply src/app/qgsoptions.cpp]: Agreed

Comment (ismailsunni) [reply src/app/qgsoptions.cpp]: Ah ok, that makes sense. I will remove it.

File: src/app/3d/qgs3dmaptoolmeasureline.cpp (hunk 1 above)
Comment (ismailsunni) [reply src/app/3d/qgs3dmaptoolmeasureline.cpp]: Yes, it's for debugging purpose only. I will remove them when it's ready.

Ah, I thought both QgsDebugMsg and qDebug are equivalent. I will use QgsDebugMsg from now then.

File: src/3d/qgs3dmapscene.cpp (hunk 15 above)
Comment (ismailsunni) [reply src/3d/qgs3dmapscene.cpp]: Nope, removed.

File: src/app/3d/qgs3dmaptoolmeasureline.cpp (hunk 5 above)
Comment (ismailsunni) [reply src/app/3d/qgs3dmaptoolmeasureline.cpp]: Done.

File: src/app/3d/qgs3dmaptoolmeasureline.cpp (hunk 2 above)
Comment (ismailsunni) [reply src/app/3d/qgs3dmaptoolmeasureline.cpp]: Done.

File: src/app/3d/qgs3dmaptoolmeasureline.cpp (hunk 4 above)
Comment (ismailsunni) [reply src/app/3d/qgs3dmaptoolmeasureline.cpp]: Done.

File: src/app/3d/qgs3dmaptoolmeasureline.h (hunk 10 above)
Comment (ismailsunni) [reply src/app/3d/qgs3dmaptoolmeasureline.h]: Ok

File: src/app/3d/qgs3dmaptoolmeasureline.cpp (hunk 3 above)
Comment (ismailsunni) [reply src/app/3d/qgs3dmaptoolmeasureline.cpp]: Done. Removed.

File: src/app/3d/qgs3dmaptoolmeasureline.cpp (hunk 9 above)
Comment (ismailsunni) [reply src/app/3d/qgs3dmaptoolmeasureline.cpp]: Done

File: src/app/3d/qgs3dmaptoolmeasureline.cpp (hunk 6 above)
Comment (ismailsunni) [reply src/app/3d/qgs3dmaptoolmeasureline.cpp]: Done

File: src/app/3d/qgs3dmaptoolmeasureline.cpp (hunk 7 above)
Comment (ismailsunni) [reply src/app/3d/qgs3dmaptoolmeasureline.cpp]: done

File: src/app/3d/qgs3dmaptoolmeasureline.cpp (hunk 8 above)
Comment (ismailsunni) [reply src/app/3d/qgs3dmaptoolmeasureline.cpp]: done.

File: src/app/3d/qgs3dmeasuredialog.cpp (hunk 11 above)
Comment (ismailsunni) [reply src/app/3d/qgs3dmeasuredialog.cpp]: done

File: src/app/3d/qgs3dmaptoolmeasureline.cpp (hunk 19 above)
Comment (ismailsunni) [reply src/app/3d/qgs3dmaptoolmeasureline.cpp]: done

File: src/app/3d/qgs3dmaptoolmeasureline.cpp (hunk 16 above)
Comment (ismailsunni) [reply src/app/3d/qgs3dmaptoolmeasureline.cpp]: done

File: src/core/qgsdistancearea.cpp (hunk 13 above)
Comment (ismailsunni) [reply src/core/qgsdistancearea.cpp]: Function is removed now.

File: python/core/auto_generated/qgsdistancearea.sip.in (hunk 14 above)
Comment (ismailsunni) [reply python/core/auto_generated/qgsdistancearea.sip.in]: Moved.

File: src/core/qgsdistancearea.cpp (hunk 13 above)
Comment (ismailsunni) [reply src/core/qgsdistancearea.cpp]: function is removed.

File: src/core/geometry/qgslinestring.cpp [hunk 21]
```diff
@@ -443,6 +443,21 @@ double QgsLineString::length() const
   return length;
 }
 
+double QgsLineString::length3D() const
+{
+  double length = 0;
```
Comment (nyalldawson) [inline src/core/geometry/qgslinestring.cpp]: If someone calls this on a LineString with no z, it'll crash. Can you add a check at the start here and if there's no z present, just return length() instead?

File: src/core/geometry/qgslinestring.h [hunk 22]
```diff
@@ -564,6 +564,7 @@ class CORE_EXPORT QgsLineString: public QgsCurve
 
     //curve interface
     double length() const override;
+    double length3D() const;
```
Comment (nyalldawson) [inline src/core/geometry/qgslinestring.h]: Doxygen

File: tests/src/core/testqgsgeometry.cpp [hunk 23]
```diff
@@ -3757,6 +3757,13 @@ void TestQgsGeometry::lineString()
   //endPoint
   QCOMPARE( l19.endPoint(), QgsPoint( QgsWkbTypes::PointZM, 15, 10, 6, 7 ) );
 
+  //test 3d length
+  l19.clear();
+  l19.setPoints( QgsPointSequence() << QgsPoint( QgsWkbTypes::PointZ, 0, 0, 0 )
+                 << QgsPoint( QgsWkbTypes::PointZ, 1, 2, 2 )
+                 << QgsPoint( QgsWkbTypes::PointZ, 4, 6, 2 ) );
+  QCOMPARE( l19.length3D(), 8.0 );
```
Comment (nyalldawson) [inline tests/src/core/testqgsgeometry.cpp]: Also:
- a unit test without z
- a LineStringZM
- an empty LineStringZ with no vertices

**Review by nyalldawson — COMMENTED** [review]
Looking good!

There's still a lot of code in common between the 2d/3d measurement dialogs. Is there any chance of refactoring these to add a common base class to avoid this?

File: src/core/geometry/qgslinestring.h (hunk 22 above)
Comment (ismailsunni) [reply src/core/geometry/qgslinestring.h]: Done.

File: tests/src/core/testqgsgeometry.cpp (hunk 23 above)
Comment (ismailsunni) [reply tests/src/core/testqgsgeometry.cpp]: Added.

File: src/core/geometry/qgslinestring.cpp (hunk 21 above)
Comment (ismailsunni) [reply src/core/geometry/qgslinestring.cpp]: Added.

**Comment by ismailsunni** [comment]
Hi @nyalldawson 

Thanks for the review.

> There's still a lot of code in common between the 2d/3d measurement dialogs. Is there any chance of refactoring these to add a common base class to avoid this?

The common base class will mostly have a few same members and 2 methods that have the same implementation since I do not use `QgsMeasureDistance` (which is used a lot in 2D). Even the tool itself (`QgsMeasureTool` and `Qgs3DMapToolMeasureLine` do not have a common base class). I did it for my first approach actually, but it was quite complicated.

File: src/app/3d/qgs3dmaptoolmeasureline.cpp [hunk 24]
```diff
@@ -0,0 +1,269 @@
...
+  {
+    restart();
+    updateSettings();
+  }
+  else
+  {
+    QgsLineString *measurementLine = new QgsLineString();
+    measurementLine->addZValue();
+
+    QgsFeature *measurementFeature = new QgsFeature( QgsFeatureId( 1 ) );
```
Comment (wonder-sk) [inline src/app/3d/qgs3dmaptoolmeasureline.cpp]: no need to do `QgsFeature *f = new QgsFeature(...);`  - let's create it on the stack: `QgsFeature f(...);`

this way you don't need to call `delete f` at the end (the current code contains a leak).

File: src/app/3d/qgs3dmaptoolmeasureline.cpp [hunk 25]
```diff
@@ -0,0 +1,269 @@
...
+
+    // Add feature to layer
+    mMeasurementLayer->startEditing();
+    mMeasurementLayer->addFeature( *measurementFeature );
+    mMeasurementLayer->commitChanges();
+
+    // Set style
+    updateSettings();
+    mIsAlreadyActivated = true;
+//    updateSettings();
```
Comment (wonder-sk) [inline src/app/3d/qgs3dmaptoolmeasureline.cpp]: best to get rid of unneeded commented code...

File: src/app/3d/qgs3dmaptoolmeasureline.cpp [hunk 26]
```diff
@@ -0,0 +1,269 @@
...
+        QgsPoint( it->x(), it->y(), it->z() / verticalScale )
+      );
+    }
+    line = new QgsLineString( descaledPoints );
+  }
+  else
+  {
+    line = new QgsLineString( mPoints );
+  }
+  QgsGeometry *lineGeometry = new QgsGeometry( line );
```
Comment (wonder-sk) [inline src/app/3d/qgs3dmaptoolmeasureline.cpp]: Also here you can just use a local variable `QgsGeometry lineGeometry( line );` - and avoid a memory leak. (the geometry map nor the changeGeometryValues() function do not take ownership of the passed geometry - they just make a copy)

File: src/app/3d/qgs3dmaptoolmeasureline.cpp [hunk 27]
```diff
@@ -0,0 +1,269 @@
...
+
+  QgsPhongMaterialSettings phongMaterial;
+  QgsSettings settings;
+  int myRed = settings.value( QStringLiteral( "qgis/default_measure_color_red" ), 222 ).toInt();
+  int myGreen = settings.value( QStringLiteral( "qgis/default_measure_color_green" ), 155 ).toInt();
+  int myBlue = settings.value( QStringLiteral( "qgis/default_measure_color_blue" ), 67 ).toInt();
+  phongMaterial.setAmbient( QColor( myRed, myGreen, myBlue, 100 ) );
+  lineSymbol->setMaterial( phongMaterial );
+
+  mLineSymbolRenderer = new QgsVectorLayer3DRenderer( lineSymbol );
```
Comment (wonder-sk) [inline src/app/3d/qgs3dmaptoolmeasureline.cpp]: At this point, a previous mLineSymbolRenderer object may still exist, so there may be a memory leak here. Best to use std::unique_ptr that takes care of destructing the old 3d renderer when needed.

File: src/app/3d/qgs3dmaptoolmeasureline.h [hunk 28]
```diff
@@ -0,0 +1,98 @@
...
+    void handleClick( Qt3DRender::QPickEvent *event, const QgsVector3D &worldIntersection );
+
+  private:
+    std::unique_ptr<Qgs3DMapToolMeasureLinePickHandler> mPickHandler;
+
+    friend class Qgs3DMapToolMeasureLinePickHandler;
+
+    QgsVectorLayer *mMeasurementLayer = nullptr;
+    QgsVectorLayer3DRenderer *mLineSymbolRenderer = nullptr;
+    bool mIsAlreadyActivated;
```
Comment (wonder-sk) [inline src/app/3d/qgs3dmaptoolmeasureline.h]: It's a good habit in modern c++ to initialize the member variables here (instead of in constructor) - so just add ` = false` here and you can remove that statement from the constructor

File: src/ui/qgsoptionsbase.ui [hunk 29]
```diff
@@ -332,7 +332,7 @@
        <item>
         <widget class="QStackedWidget" name="mOptionsStackedWidget">
          <property name="currentIndex">
-          <number>8</number>
+          <number>6</number>
```
Comment (wonder-sk) [inline src/ui/qgsoptionsbase.ui]: In the end there are no changes to the options dialog right? It would be nice to revert the changes in the .ui file on this PR...

**Review by wonder-sk — COMMENTED** [review]
Nearly there :-)

File: src/core/geometry/qgslinestring.h [hunk 30]
```diff
@@ -564,6 +564,13 @@ class CORE_EXPORT QgsLineString: public QgsCurve
 
     //curve interface
     double length() const override;
+
+    /**
+     * Returns the length in 3D world of the line string.
+     * If it is not a 3D line string, return its 2D length.
+     * \see length()
+     */
```
Comment (nyalldawson) [inline src/core/geometry/qgslinestring.h]: \since QGIS 3.10

File: src/app/3d/qgs3dmaptoolmeasureline.h (hunk 28 above)
Comment (ismailsunni) [reply src/app/3d/qgs3dmaptoolmeasureline.h]: Hmm, I didn't know about it. Added. Thanks.

File: src/app/3d/qgs3dmaptoolmeasureline.cpp (hunk 25 above)
Comment (ismailsunni) [reply src/app/3d/qgs3dmaptoolmeasureline.cpp]: Removed.

File: src/app/3d/qgs3dmaptoolmeasureline.cpp (hunk 24 above)
Comment (ismailsunni) [reply src/app/3d/qgs3dmaptoolmeasureline.cpp]: Done

File: src/app/3d/qgs3dmaptoolmeasureline.cpp (hunk 26 above)
Comment (ismailsunni) [reply src/app/3d/qgs3dmaptoolmeasureline.cpp]: Done.

File: src/ui/qgsoptionsbase.ui (hunk 29 above)
Comment (ismailsunni) [reply src/ui/qgsoptionsbase.ui]: Done.

File: src/app/3d/qgs3dmaptoolmeasureline.cpp (hunk 27 above)
Comment (ismailsunni) [reply src/app/3d/qgs3dmaptoolmeasureline.cpp]: Hi @wonder-sk 
Actually, I pass that `mLineSymbolRenderer` to the `mMeasurementLayer` in the `setMeasurementLayerRenderer` function. It set by `mMeasurementLayer->setRenderer3D( lineSymbolRenderer );`, I check in the code, whenever it calls `setRenderer3D`, it will delete the old `3DRenderer`.

So, it's safe to use `new` here, isn't it?

Btw, if I use std:unique_ptr, it will crash whenever I call the `setMeasurementLayerRenderer`. Probably because it is pointed by the layer now.

Comment (wonder-sk) [reply src/app/3d/qgs3dmaptoolmeasureline.cpp]: true - sorry my bad, I have forgotten the ownership is taken by the layer...

**Review by wonder-sk — APPROVED** [review]
